app/scripts/controller-init/types.ts

```diff
@@ -110,6 +110,16 @@ export type ControllerInitRequest<
    * e.g. `{ TransactionController: { transactions: [] } }`.
    */
   persistedState: ControllerPersistedState;
+
+  /**
+   * The `MetaMaskController` instance.
+   *
+   * This should only be used if the controller messenger cannot be used
+   * instead.
+   */
+  // `MetaMaskController` is untyped.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  metaMaskController: any;
```

Comment:
I added this as temporary workaround for accessing the `getPrimaryMnemonic` and `removeAllConnections` functions of `MetaMaskController`. These are required for initialisation of the `SnapController`.

Alternatively we can create a (typed) messenger for the `MetaMaskController` which exposes these functions.

Reply:
I think the long-term cleanest fix is to modularise and isolate any client functionality we need in a suitable messenger such as one for connections specifically.

In the interest of a quick fix, I definitely think we want to avoid coupling to the entire `MetamaskController` as that's precisely what we're trying to move away from as it still contains all the legacy controller instances.

I would vote for the safest "quick" fix to be:

- Get the primary key via `request.getController('KeyringController').getKeyringsByType`.
  - Assuming there isn't an existing messenger action we could use instead.
- Add only `removeAllConnections` itself to the request to vastly limit the coupling

Reply:
Updated in aea013c.